### PR TITLE
feat: make query be aware of timezone setting

### PR DESCRIPTION
**Pull request by killme2008**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Try to aware of timezone setting for `TypeConversionRule`, changes:

* Added `ExtensionAnalyzerRule` that extends DataFusion `AnalyzerRule` to accept `QueryEngineContext` as the second parameter.
* Implement `ExtensionAnalyzerRule` for `TypeConversionRule`.
* Optimize by extension analyzer rules at first in `DfLogicalPlanner` and `LogicalOptimizer`.
* Fixed MySQL `show variables time_zone` can't work.
* Adds timezone sqlness test.
* Make `Date`, `Timestamp` and `DateTime` parsing from string value aware of timezone.
* Make UDF functions aware of timezone.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.
- [ ]  This PR does not require documentation updates.

## Refer to a related PR or issue link (optional)

#2907

### CODE REVIEW

File: src/query/src/optimizer.rs
Comment (v0y4g3r) [inline src/query/src/optimizer.rs:24]: `AnalyzerRuleExtension` would be better. It's an extension to analyzer rules, not analyzer rules to extensions.

File: src/sql/src/statements.rs
Comment (v0y4g3r) [inline src/sql/src/statements.rs:97]: Also respect per-session time zone settings here?

File: src/datatypes/src/value.rs
Comment (killme2008) [reply src/datatypes/src/value.rs:434]: Ok, I'll address all these issues in the next PR.

File: src/query/src/optimizer.rs
Comment (killme2008) [reply src/query/src/optimizer.rs:24]: Just keep the same as `DistExtensionPlanner`.

Other cases such as `substrait::extension_serializer::ExtensionSerializer`.

File: src/datatypes/src/value.rs
Comment (killme2008) [reply src/datatypes/src/value.rs:434]: @v0y4g3r  Looks like it's unreachable here, the `TypeConversionRule` will convert all the `ScalarValue::Utf8` to `ScalarValue::Timestamp`s in the timestamp predicate case.

File: src/query/src/optimizer.rs
Comment (killme2008) [reply src/query/src/optimizer.rs:29]: Yes, it's unused currently. I'll remove it.

File: src/common/time/src/util.rs
Comment (killme2008) [reply src/common/time/src/util.rs:23]: Of course, leave it to the following PRs.

File: src/sql/src/statements.rs
Comment (killme2008) [reply src/sql/src/statements.rs:97]: Leave it to the other PR.

**Comment by killme2008** [comment]
I'll rebase https://github.com/GreptimeTeam/greptimedb/pull/3193 after it is merged.

File: src/common/time/src/util.rs
Comment (killme2008) [reply src/common/time/src/util.rs:23]: Looks like we don't need to change it. The function is to format the naive utc datetime into string by the pattern.

**Comment by killme2008** [comment]
@waynexia @v0y4g3r Please take another look. Already made `Date` and `DateTime` parsing from string value be aware of timezone.

I think all the queries will be aware of the time zone right now.

File: tests/cases/standalone/common/system/timezone.sql
Comment (waynexia) [inline tests/cases/standalone/common/system/timezone.sql:25]: I just realized PG's timezone is different from other variables: https://www.postgresql.org/docs/current/sql-set.html

```sql
SET TIME ZONE '+08:00';
SET TIME ZONE 'Asia/Shanghai';
```

Maybe we need another PR to align this behavior

File: tests/cases/standalone/common/system/timezone.sql
Comment (waynexia) [inline tests/cases/standalone/common/system/timezone.sql:1]: sugg: Consider adding some time string with timezone, like a canonical rfc3339 format.

File: src/sql/src/statements.rs
Comment (waynexia) [inline src/sql/src/statements.rs:87]: Why one is `from_str_utc` and another is `from_str_local`?

File: src/script/src/python/engine.rs
Comment (waynexia) [inline src/script/src/python/engine.rs:174]: Why not `with_context()` here? Formatted error losses it context

File: src/query/src/query_engine/state.rs
Comment (waynexia) [inline src/query/src/query_engine/state.rs:136]: One remaining `&mut`

File: src/cmd/src/cli/repl.rs
Comment (waynexia) [inline src/cmd/src/cli/repl.rs:172]: another remaining `&mut`

File: src/sql/src/statements.rs
Comment (killme2008) [reply src/sql/src/statements.rs:87]: Because the old DateTime `from_str` implementation uses the local timezone when the string doesn't contain any timezone info. I keep it the same way.


https://github.com/GreptimeTeam/greptimedb/blob/966875ee1120bc0189d6303d4ea8fb445cdcf7a7/src/common/time/src/datetime.rs#L69

**Comment by killme2008** [comment]
@waynexia PTAL

File: src/common/time/src/timestamp.rs
Comment (v0y4g3r) [inline src/common/time/src/timestamp.rs:435]: ditto

File: src/query/src/parser.rs
Comment (v0y4g3r) [inline src/query/src/parser.rs:171]: Is this todo intentionally removed so that promql timestamp literal parsing does not consider time zone?

**Review by v0y4g3r — APPROVED** [review]
Generally LGTM

File: src/sql/src/statements.rs
Comment (waynexia) [reply src/sql/src/statements.rs:87]: >when the timezone is not provided both in string and parameter.

We should then fallback to server timezone, then UTC+0. Local timezone isn't one of the options

**Review by waynexia — APPROVED** [review]
Others LGTM, except the from/to local tz

File: src/sql/src/statements.rs
Comment (killme2008) [reply src/sql/src/statements.rs:87]: Local timezone is the system timezone.

File: src/query/src/parser.rs
Comment (killme2008) [reply src/query/src/parser.rs:171]: The `rfc3339` format contains the timezone info, so we don't need to this fix.

File: src/common/time/src/timestamp.rs
Comment (killme2008) [reply src/common/time/src/timestamp.rs]: Addressed in https://github.com/GreptimeTeam/greptimedb/pull/3175/commits/7c3bd196d83f285370940e0cce125e5a3bf533a5

**Comment by killme2008** [comment]
> Others LGTM, except the from/to local tz

Renamed the `local` to `system` in https://github.com/GreptimeTeam/greptimedb/pull/3175/commits/7c3bd196d83f285370940e0cce125e5a3bf533a5

**Pull request merged**
